### DISCUSSION TITLE
How to get F1 Score, Precision, Recall and Accuracy after testing

### DISCUSSION THREAD

**Discussion by Zidny000** [ultralytics/ultralytics#16196] [General]
I am trying to get the F1 Score, Precision, Recall, and Accuracy after testing the model on a dataset but I am having hard to find a correct way to do that I was about to generate a result.csv file inside val folder but the data isn't labeled so I am not understanding anything. I am attaching the csv file if anyone can help me with this. 
[test_val_results.csv](https://github.com/user-attachments/files/16948433/test_val_results.csv)

**Comment by glenn-jocher** [comment]
@Zidny000 to obtain F1 Score, Precision, Recall, and Accuracy, ensure your dataset is labeled correctly. Use the validation mode in YOLOv8, which provides these metrics automatically. For more details, refer to the model validation documentation.

**Reply by Zidny000** [reply]
@glenn-jocher thanks for your responnse
!yolo task=segment mode=val model=/path/best.pt data={dataset}/data.yaml iou=0.5 imgsz=640 > /path/test_val_results.csv
This is the command I am using to test the model. The dataset is correctly labeled and optimized for YOLOv8. 
![val folder](https://github.com/user-attachments/assets/e399655d-cfa0-4a10-a0cf-345764f6758a)

this are the files I am getting in my val folder. Can you please suggest how I can get the below metrics?
Precision, Recall (Sensitivity), Specificity, Accuracy, F1 Score, IoU and a confusion matrix in pixel level

**Reply by glenn-jocher** [reply]
To obtain these metrics, ensure your validation command is correct. YOLOv8's validation mode should automatically provide Precision, Recall, F1 Score, and IoU. For pixel-level metrics and a confusion matrix, consider post-processing the results using a custom script.

**Reply by Zidny000** [reply]
I have searched for this a lot in the documentation but didn't find any solution or specific command for validation to get these results. I am providing my code if you can suggest me any approach or what could be the right command. Please have a look at it. I am using the model for medical image segmentation.
[YOLOv8.txt](https://github.com/user-attachments/files/17152616/YOLOv8.txt)

**Reply by glenn-jocher** [reply]
@Zidny000 for detailed metrics like Precision, Recall, and F1 Score, ensure your dataset is correctly labeled and use the validation mode. For pixel-level metrics, consider using a custom script to analyze the results further. If you need specific guidance, please share more details about your setup.

**Reply by Zidny000** [reply]
I am segmenting the Small Cell Lung Cancer from lung CT, PET scans. The scans were in DICOM format and annotated  when I took them from TCIA. After that I have manually segmented the ROI using roboflow from the scans to train and test the model. I am providing the zip file for the test dataset please have a look at it.
[test_dataset.zip](https://github.com/user-attachments/files/17161325/test_dataset.zip)

**Reply by glenn-jocher** [reply]
I'm unable to access files directly. Please ensure your dataset is correctly formatted for YOLOv8. If you encounter specific issues, feel free to describe them here for further assistance.

**Reply by Zidny000** [reply]
If you can't access the files then let me give one scan png example and its corresponding label
![B0031_1-25_png_jpg rf 7b8b4d327fe70c35055ce2d5e0ad9709](https://github.com/user-attachments/assets/fce83aae-71c3-4277-adfe-6dd48418e341)
The label for this image is below which is a .txt file
0 0.63232421875 0.52978515625 0.62744140625 0.5517578125 0.62744140625 0.57861328125 0.6298828125 0.59814453125 0.654296875 0.6103515625 0.63720703125 0.6103515625 0.61279296875 0.60791015625 0.634765625 0.6298828125 0.63232421875 0.64453125 0.62744140625 0.6689453125 0.63720703125 0.68359375 0.66650390625 0.68359375 0.68603515625 0.6494140625 0.68603515625 0.6298828125 0.6982421875 0.60791015625 0.7275390625 0.5859375 0.72509765625 0.56396484375 0.7177734375 0.537109375 0.68115234375 0.5126953125 0.63232421875 0.52978515625

data.yaml file
names:
- Segmentation Mask - v1 2024-04-08 12-41am
nc: 1
roboflow:
  license: Private
  project: sclc-segmentation-mask-640
  url: https://app.roboflow.com/data-gw6dx/sclc-segmentation-mask-640/1
  version: 1
  workspace: data-gw6dx
test: ./test/images
train: ./train/images
val: ./test/images

My data contains only one class which is SCLC(Small Cell Lung Cancer) represented by 0 in the label file. And the dataset was prepared using roboflow for specifically YOLOv8 format. 
But after validating the model I don't get any metrics in the val folder. 
Inside cfg folder of ultralytics where I have changed below parameters to true in hope of getting the metrics
save_json: True # (bool) save results to JSON file
save_txt: True # (bool) save results as .txt file

I do get a protection.json file after this in the val folder but it does not come handy to extract the required metrics

**Reply by glenn-jocher** [reply]
To obtain the metrics, ensure your validation command is set up correctly and that your dataset is properly formatted. If the metrics aren't appearing, try running the validation again and check the console output for any errors. If issues persist, consider using a custom script to calculate the metrics from the predictions.

**Reply by Zidny000** [reply]
The console output has no errors 
!yolo task=segment mode=val save_txt=True save_conf=True save_crop=True model=/path/best.pt data={dataset}/data.yaml imgsz=640

This is the command I am using to run the validation.
As you suggested write a custom script. Can you tell me where to look for it in the code or which file contains the code for metrics calculation? Any kind of suggestions or documentation will be helpful. Thank you.

Also below is the console output if it helps to figure out something

/content/drive/MyDrive/Models/YOLOv8/Metrics/YOLOv8/ultralytics/ultralytics/nn/tasks.py:732: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  ckpt = torch.load(file, map_location="cpu")
Ultralytics YOLOv8.2.2 🚀 Python-3.10.12 torch-2.4.1+cu121 CUDA:0 (Tesla T4, 15102MiB)
YOLOv8x-seg summary (fused): 295 layers, 71721619 parameters, 0 gradients, 343.7 GFLOPs
val: Scanning /content/drive/MyDrive/Models/YOLOv8/Metrics/YOLOv8/custom_dataset/test/labels.cache... 131 images, 0 backgrounds, 0 corrupt: 100% 131/131 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95): 100% 9/9 [00:15<00:00,  1.73s/it]
                   all        131        131      0.625       0.56      0.605      0.401       0.63      0.557      0.588      0.373
Speed: 1.2ms preprocess, 71.9ms inference, 0.0ms loss, 14.9ms postprocess per image
Saving runs/segment/val11/predictions.json...
Results saved to runs/segment/val11

**Reply by glenn-jocher** [reply]
To calculate additional metrics, you can modify the validation script in the Ultralytics codebase. Look into the `val.py` file for metric calculations. For custom metrics, consider using libraries like scikit-learn to process the predictions saved in `predictions.json`.
